File: onco/public/js/landed_cost_voucher.js
frappe.ui.form.on('Landed Cost Voucher', {
    onload: function(frm) {
        frm.set_query('receipt_document', 'purchase_receipts', function(doc, cdt, cdn) {
            let child = locals[cdt][cdn];
            if (child.receipt_document_type === 'Purchase Invoice') {
                return {
                    filters: {
                        docstatus: 1,
                        company: doc.company
                    }
                };
            }
        });
    },
    custom_shipment_id: function(frm) {
        if (frm.doc.custom_shipment_id) {
            frappe.show_alert({message: `Fetching vendor invoices for Shipment ID ${frm.doc.custom_shipment_id}...`, indicator: 'green'});
            fetch_vendor_invoices(frm, frm.doc.custom_shipment_id);
        }
    }
});

frappe.ui.form.on('Landed Cost Purchase Receipt', {
    receipt_document: function(frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.receipt_document_type === 'Purchase Invoice' && row.receipt_document) {
            frappe.call({
                method: 'onco.onco.custom_scripts.landed_cost_voucher.get_shipment_from_purchase_invoice',
                args: { purchase_invoice: row.receipt_document },
                callback: function(r) {
                    if (r && r.message) {
                        frm.set_value('custom_shipment_id', r.message);
                        fetch_vendor_invoices(frm, r.message);
                    }
                }
            });
        }
    }
});

function fetch_vendor_invoices(frm, shipment_id) {
    frappe.call({
        method: 'onco.onco.custom_scripts.landed_cost_voucher.get_vendor_invoices_for_shipment',
        args: {
            shipment_id: shipment_id,
            company: frm.doc.company
        },
        callback: function(r) {
            if (r && r.message && r.message.length > 0) {
                frm.clear_table('vendor_invoices');

                if (frm.doc.taxes && frm.doc.taxes.length > 0) {
                    let first_row = frm.doc.taxes[0];
                    if (!first_row.expense_account && !first_row.description && !first_row.amount) {
                        frm.clear_table('taxes');
                    }
                }

                let current_taxes = frm.doc.taxes || [];
                let existing_invoices = current_taxes.map(t => t.custom_vendor_invoice);
                let added_to_taxes = 0;
                let added_to_vi = 0;

                let current_vi = (frm.doc.vendor_invoices || []).map(v => v.vendor_invoice);

                r.message.forEach(invoice => {
                    if (invoice.remaining > 0 && !current_vi.includes(invoice.name)) {
                        let vi_row = frm.add_child('vendor_invoices');
                        vi_row.vendor_invoice = invoice.name;
                        vi_row.amount = invoice.remaining;
                        vi_row.custom_reference_shipment = invoice.reference_shipment;
                        added_to_vi++;
                    }

                    if (!existing_invoices.includes(invoice.name)) {
                        let child = frm.add_child('taxes');
                        child.description = invoice.description;
                        child.expense_account = invoice.expense_account;
                        child.amount = invoice.remaining;
                        child.custom_vendor_invoice = invoice.name;
                        child.custom_supplier_invoice_no = invoice.bill_no;
                        child.custom_posting_date = invoice.posting_date;
                        child.custom_allocated_to_shipment = invoice.allocated_to_shipment;
                        child.custom_already_used = invoice.already_used;
                        child.custom_remaining = invoice.remaining;
                        added_to_taxes++;
                    }
                });

                frm.refresh_field('vendor_invoices');
                frm.refresh_field('taxes');

                if (added_to_vi > 0) {
                    frappe.show_alert({message: `Added ${added_to_vi} vendor invoice(s) and ${added_to_taxes} tax row(s).`, indicator: 'green'});
                }

                let exhausted = r.message.filter(inv => inv.remaining <= 0);
                if (exhausted.length > 0) {
                    frappe.msgprint({
                        title: 'Notice',
                        indicator: 'orange',
                        message: `${exhausted.length} invoice(s) have no remaining balance to allocate. They were skipped.`
                    });
                }

                if (frm.doc.purchase_receipts && frm.doc.purchase_receipts.length > 0) {
                    setTimeout(() => {
                        frappe.call({
                            method: "get_items_from_purchase_receipts",
                            doc: frm.doc,
                            callback: function(r2) {
                                if (!r2.exc) {
                                    if (frm.doc.items && frm.doc.items.length > 0 && !frm.doc.items[0].item_code) {
                                        frm.doc.items.shift();
                                    }
                                    frm.refresh_field("items");
                                }
                            }
                        });
                    }, 500);
                }
            } else {
                frappe.msgprint(`No submitted vendor service invoices found for Shipment ID: ${shipment_id}`);
            }
        }
    });
}

File: onco/onco/custom_scripts/landed_cost_voucher.py
# Copyright (c) 2026, Onco and contributors

"""
Landed Cost Voucher Customizations
- Auto-fetch vendor invoices based on Shipment ID (multi-shipment table)
- Cumulative tracking (allocated, already used, remaining)
- Vendor invoice details in taxes table
"""

import frappe
from frappe import _
from frappe.utils import flt


@frappe.whitelist()
def get_vendor_invoices_for_shipment(shipment_id, company):
    """
    Fetch all vendor service Purchase Invoices linked to a shipment
    via the Shipment Allocation Reference child table, with cumulative tracking.

    Args:
        shipment_id: The Shipment document name
        company: Company name for filtering

    Returns:
        List of vendor invoices with allocation & cumulative details
    """
    if not shipment_id:
        frappe.throw(_("Shipment ID is required"))

    vendor_invoices = frappe.db.sql("""
        SELECT
            pi.name,
            pi.supplier,
            pi.supplier_name,
            pi.posting_date,
            pi.grand_total,
            pi.currency,
            pi.bill_no,
            pi.remarks,
            sar.amount as allocated_to_shipment,
            sar.shipment_id as reference_shipment
        FROM
            `tabPurchase Invoice` pi
        INNER JOIN
            `tabShipment Allocation Reference` sar
            ON sar.parent = pi.name
            AND sar.parenttype = 'Purchase Invoice'
            AND sar.parentfield = 'custom_shipment_allocation'
        WHERE
            pi.name IN (
                SELECT DISTINCT sar2.parent
                FROM `tabShipment Allocation Reference` sar2
                WHERE sar2.shipment_id = %(shipment_id)s
                    AND sar2.parenttype = 'Purchase Invoice'
                    AND sar2.parentfield = 'custom_shipment_allocation'
            )
            AND pi.docstatus = 1
            AND pi.company = %(company)s
        ORDER BY
            pi.name, sar.shipment_id, sar.idx
    """, {
        'shipment_id': shipment_id,
        'company': company
    }, as_dict=True)

    if not vendor_invoices:
        return []

    result = []
    for invoice in vendor_invoices:
        expense_account = get_primary_expense_account(invoice.name)

        already_used = get_already_used_in_lcv(invoice.name, invoice.reference_shipment)
        remaining = invoice.allocated_to_shipment - already_used
        if remaining < 0:
            remaining = 0

        result.append({
            'name': invoice.name,
            'supplier': invoice.supplier,
            'supplier_name': invoice.supplier_name,
            'posting_date': invoice.posting_date,
            'grand_total': invoice.grand_total,
            'currency': invoice.currency,
            'bill_no': invoice.bill_no or '',
            'reference_shipment': invoice.reference_shipment,
            'expense_account': expense_account,
            'allocated_to_shipment': invoice.allocated_to_shipment,
            'already_used': already_used,
            'remaining': remaining,
            'description': f"{invoice.supplier_name} - {invoice.name}"
                + (f" ({invoice.remarks})" if invoice.remarks else "")
        })

    return result


def get_already_used_in_lcv(purchase_invoice_name, shipment_id):
    """
    Calculate how much of a vendor invoice's allocation to a specific shipment
    has already been consumed by submitted Landed Cost Vouchers.

    Args:
        purchase_invoice_name: The vendor Purchase Invoice name
        shipment_id: The Shipment ID

    Returns:
        Total amount already used in submitted LCVs
    """
    result = frappe.db.sql("""
        SELECT COALESCE(SUM(tax.amount), 0) as used
        FROM `tabLanded Cost Taxes and Charges` tax
        INNER JOIN `tabLanded Cost Voucher` lcv
            ON tax.parent = lcv.name
        WHERE
            tax.custom_vendor_invoice = %(pi)s
            AND lcv.custom_shipment_id = %(shipment)s
            AND lcv.docstatus = 1
    """, {
        'pi': purchase_invoice_name,
        'shipment': shipment_id
    }, as_dict=True)

    return result[0].used if result else 0


def get_primary_expense_account(purchase_invoice_name):
    """
    Get the primary expense account from a Purchase Invoice
    Uses the most common expense account from invoice items
    """
    accounts = frappe.db.sql("""
        SELECT
            expense_account,
            COUNT(*) as count,
            SUM(amount) as total_amount
        FROM
            `tabPurchase Invoice Item`
        WHERE
            parent = %(invoice)s
            AND expense_account IS NOT NULL
        GROUP BY
            expense_account
        ORDER BY
            total_amount DESC
        LIMIT 1
    """, {
        'invoice': purchase_invoice_name
    }, as_dict=True)

    if accounts and len(accounts) > 0:
        return accounts[0].expense_account

    default_account = frappe.db.get_value(
        'Purchase Invoice',
        purchase_invoice_name,
        'against_expense_account'
    )

    if default_account:
        return default_account.split(',')[0].strip()

    company = frappe.db.get_value('Purchase Invoice', purchase_invoice_name, 'company')
    default_expense = frappe.db.get_value(
        'Company',
        company,
        'default_expense_account'
    )

    return default_expense or ''


def validate_landed_cost_voucher(doc, method):
    """
    Validation hook for Landed Cost Voucher
    Ensures all linked PIs reference the same shipment via the allocation table
    """
    if doc.custom_shipment_id and doc.purchase_receipts:
        for voucher_row in doc.purchase_receipts:
            if voucher_row.receipt_document and voucher_row.receipt_document_type == 'Purchase Invoice':
                pi_shipments = frappe.db.sql("""
                    SELECT shipment_id
                    FROM `tabShipment Allocation Reference`
                    WHERE parent = %(pi)s
                        AND parenttype = 'Purchase Invoice'
                        AND parentfield = 'custom_shipment_allocation'
                """, {'pi': voucher_row.receipt_document}, as_dict=True)

                if pi_shipments:
                    pi_shipment_ids = [s.shipment_id for s in pi_shipments]
                    if doc.custom_shipment_id not in pi_shipment_ids:
                        frappe.throw(_(
                            "Purchase Invoice {0} is allocated to shipments {1}, "
                            "but this Landed Cost Voucher is for Shipment {2}"
                        ).format(
                            voucher_row.receipt_document,
                            ', '.join(pi_shipment_ids),
                            doc.custom_shipment_id
                        ))

                legacy_shipment = frappe.db.get_value(
                    'Purchase Invoice',
                    voucher_row.receipt_document,
                    'custom_shipments'
                )

                if legacy_shipment and legacy_shipment != doc.custom_shipment_id:
                    frappe.throw(_(
                        "Purchase Invoice {0} belongs to Shipment {1} (goods shipment), "
                        "but this Landed Cost Voucher is for Shipment {2}"
                    ).format(voucher_row.receipt_document, legacy_shipment, doc.custom_shipment_id))

    # Validate that the cumulative landed cost applied to a vendor invoice
    # (already used in submitted LCVs + this voucher) never exceeds the
    # vendor invoice's total amount.
    for tax in (doc.taxes or []):
        if not tax.custom_vendor_invoice:
            continue

        invoice_total = frappe.db.get_value(
            "Purchase Invoice", tax.custom_vendor_invoice, "grand_total"
        ) or 0
        already_used = get_already_used_in_lcv(
            tax.custom_vendor_invoice, doc.custom_shipment_id
        )
        applied_total = flt(tax.amount) + flt(already_used)

        if applied_total > flt(invoice_total):
            frappe.throw(_(
                "Landed cost for vendor invoice {0} exceeds its total amount {1}. "
                "Already used in other Landed Cost Vouchers: {2}, this voucher: {3}."
            ).format(
                tax.custom_vendor_invoice,
                invoice_total,
                already_used,
                tax.amount
            ))

def before_submit_landed_cost_voucher(doc, method):
    """
    Before submit hook for Landed Cost Voucher
    Log the landed cost application on the Shipment document
    """
    if doc.custom_shipment_id:
        try:
            shipment_doc = frappe.get_doc('Shipments', doc.custom_shipment_id)
            shipment_doc.add_comment(
                'Info',
                f'Landed Cost Voucher {doc.name} applied with total charges: {doc.total_taxes_and_charges}'
            )
        except Exception as e:
            frappe.log_error(f"Could not add comment to Shipment: {str(e)}")


@frappe.whitelist()
def get_shipment_from_purchase_invoice(purchase_invoice):
    """
    Get the first Shipment ID from a Purchase Invoice's allocation table.
    Returns the first shipment found in Shipment Allocation Reference,
    or the legacy custom_shipments field as fallback.
    """
    if not purchase_invoice:
        return None

    shipments = frappe.db.sql("""
        SELECT shipment_id
        FROM `tabShipment Allocation Reference`
        WHERE parent = %(pi)s
            AND parenttype = 'Purchase Invoice'
            AND parentfield = 'custom_shipment_allocation'
        ORDER BY idx ASC
        LIMIT 1
    """, {'pi': purchase_invoice}, as_dict=True)

    if shipments:
        return shipments[0].shipment_id

    legacy = frappe.db.get_value('Purchase Invoice', purchase_invoice, 'custom_shipments')
    return legacy or None
